Migrate header items to slot fills

joshuatf opened the pull request:
Fixes (part of) #7689 

Migrates header items to slot fills to keep feature logic separate and allow header extensibility.

Follow-ups/thoughts:

* We need to export the header slots for use externally.  GB has very specific packages for each area, but this may be overkill for our use case (i.e. `@woocommerce/header`).  Depending on how much we anticipate adding, we could do create a `@woocommerce/layout` package.  Up for discussion.
* It seems that SlotFill does not support any ordering 😱 .  This PR uses the flexbox `order` property to organize items, but I really wish there was an official API or method for sorting these fills.
* The back button could have more generalized use, but it does not at the moment and is specific to tasks so I've moved it there.  I did make the component a bit more generic though.  Maybe this should be moved back into the header and we should create a `@woocommerce/header` package?

### Screenshots

![Screen Shot 2021-10-15 at 9 21 51 AM](https://user-images.githubusercontent.com/10561050/137495482-a7579189-0238-43ea-afd4-d848a83d3b74.png)

### Detailed test instructions:

1. Navigate to various pages and make sure the activity panel works as expected and still shows the correct items.
2. Enable the navigation feature under WooCommerce->Settings->Advanced->Features.
3. Navigate to embed and non-embed pages, checking that the navigation works as expected.
4. Navigate into various tasks and make sure the task title is shown correctly and that the back button is rendered in the header.
5. Either spoof your browser as Android or change the `platform() === ANDROID_PLATFORM` manually to `true`.
6. Delete the `android_app_banner_dismissed` user meta option.
7. Navigate to any WCA page and narrow the browser to <782px, making sure the banner is shown, positioned correctly, and works as expected.

joshuatf committed:
Add SlotFill area to header

joshuatf committed:
Add activity panel fill

joshuatf committed:
Move activity panel to root client folder

joshuatf committed:
Move activity panel registration to its own folder

joshuatf committed:
Add navigation fill

joshuatf committed:
Add page title slotfill

joshuatf committed:
Slot fill the back button

joshuatf committed:
Move mobile banner to slot fill

joshuatf committed:
Fix navigation fill on embed pages

joshuatf committed:
Add changelog entry

louwie17 reviewed (commented):
Thanks for working on this @joshuatf, this tested quite well, I just ran into one issue with the Android banner. I did leave several thoughts about these changes below, let me know what you think? I think there is some room for discussion on some of these changes.

I like how you moved the `activity-panel` up one directory level, although this might cause some confusion with the activity panel under the `homescreen` folder. So maybe we want to rename it to `header-activity-panels`? As combining the folders doesn't seem like the right approach.

I also am not overly set on the `WooHeaderPageTitle` slot-fill and I think we should be more specific with the `WooHeaderItem` slot-fill.
The `WooHeaderPageTitle` slot-fill kinda relates to the back button css ordering trick. In my opinion we should keep the beginning of the header bar pretty strict, given it acts as our navigation and we don't necessarily want anyone to plop any component in there. I think using react-router for handling this would actually be better, or just keeping it the way it is now (we should just make use of the data store for the task titles).

We can keep the `WooHeaderItem` slot-fill relatively the same, maybe wrapping it in a div that more strictly aligns to the right of the header. Making it more activity-panel specific. Meaning we might want to add the `inbox` and `help` panels as separate fills?

> The back button could have more generalized use, but it does not at the moment and is specific to tasks so I've moved it there. I did make the component a bit more generic though. Maybe this should be moved back into the header and we should create a @woocommerce/header package?

Yeah I like this idea better then slot filling the back button and header title. This would allow us to have a generic header component and a generic back button component (+ more), and other plugins could choose to build their own header with our exported components or use our default less customizable one (if they are registering their own page).

louwie17 commented:
> Activity widgets maybe?

Activity widgets it is! It rolls off the tongue a little strange, but that doesn't matter to much.

> The new Woo navigation which needs to be left of everything. And the back button, which could potentially be a separate slotfill if necessary (though I'm not sure if that's overcomplicating things).

Gotcha thanks, yeah a separate slot-fill for the back button would be to much.

> No, the header is not optional the way things are currently set up. If you register a page in WCA, you get the entire layout (currently without much flexibility).

Gotcha, then the slot fills would be more helpful in this regard. I guess I would still be in favour of doing the entire header component, but I am not dead set on that. So we could keep it the way you have currently.

> I can see an argument being made to not allow SlotFill'ing the page title if we force all pages to become actual registered pages, but we break the mold on this one as well with tasks. Since tasks are not registered pages, we would have no way of updating the page title text in individual tasks. I'm very opposed to keeping random page logic in the header like we had it.

Yeah I was thinking we could turn tasks into JS router pages, and just create this config once the tasks are loaded, but this would be overkill, given we already have something that is working well in place.

> I see (and maybe incorrectly so) SlotFills as our JS-based approach to filters and a relatively low complexity solution. Even WP allows filtering of the_title so I'm curious why you feel this is a complex solution.

I am thinking more complex, because it will be trickier to debug or figure out where the title is coming from (for example). The same argument could be made for the `apply_filters`, but I suppose that is just the nature of extending to other plugins.

> This would imply that our pages are hierarchical, which they're not and this would be quite a complex refactor to make this happen. I would love it if the back button were able to detect the parent page, but this doesn't seem very feasible at the moment, especially on tasks which is the only place we currently use this and aren't even actual registered pages.

Yeah I guess slot filling or keeping the original logic would be the best way.
Did you try looking into the sorting of `fills` a little more? I just tried to find a hacky way, but couldn't come up with anything good, that part is really too bad. 

> I'm not opposed to a data store solution, but how would the task titles end up in the data store? Seems like we'd need to make use of applyFilters which we've discouraged use of.

I feel like we are referring to two different things here, I was thinking the task titles come from the registered task's title (aka the store). An external plugin would use this as well wouldn't they to register a new task?

It seems like you have slowly changed my mind @joshuatf, I am warming up to your above solution now. I think the only one that might have to be in a separate slot-fill would be the Android Banner.
I am off all of next week, so an extra opinion/set of eyes would be great on this.

louwie17 commented:
> This kind of seems like a no-brainer to me, but I'm happy to discuss layout or additional slots if we think the proposed slot here is not adequate for the time being.

@joshuatf I have been poking around with this for a little while longer and propose a slightly different slot-fill design, which could potentially go two ways, but will give a cleaner API for the sorting. It could be a semi long explanation:

Both the `Fill` and `Slot` components allow for functions as children that have access to the `children`, which gives us access to the children props. This would allow for a very basic sorting example like this:
```
const OrderedHeaderItem = ( { children } ) => children;

const WooHeaderItem = ( { children, order = 1 } ) => {
	return (
		<Fill name={ 'woocommerce_header_item' }>
		  <OrderedHeaderItem order={ order }>
		     { children }
		  </OrderedHeaderItem>
		</Fill>
	);
}

WooHeaderItem.Slot = ( { fillProps } ) => (
	<Slot name={ 'woocommerce_header_item' } fillProps={ fillProps }>
		{ ( fills ) => {
		        // sorting the fills by the order prop
			return fills.sort( ( a, b ) => {
                                 // the first child of a fill will always be OrderedHeaderItem with a order prop, as we control the WooHeaderItem
				return a[ 0 ].props.order - b[ 0 ].props.order;
			} );
		} }
	</Slot>
);
```
We could just use this, but this still doesn't play well with the HeaderPageTitle, so my proposal is this:
Have a `WooHeader` slot fill component that can contain multiple wrapper components of either a `WooHeaderPageTitle` or a `WooHeaderItem`, the `WooHeaderItem` takes an optional order, the header title will default to `0` as the order.
We did only want to decide if we merge all the fills children, or only render the last one.
Here is some potential sudo code for this:
```
const WooHeaderItem = ( { children, order } ) => children;
const WooHeaderPageTitle = ( { children } ) => children;

const WooHeader = ( { children } ) => {
	return (
		<Fill name={ 'woocommerce_header_item' }>
		  { ( props ) => {
		       if ( isFunction( children ) ) {
		          children = children( props );
	               }
		       return Children.map(children, (child, childIndex ) => {
				if ( child.type && child.type.name === 'WooHeaderPageTitle' ) {
					return (
						<WooHeaderItem order={ 0 }>
                                                          // Our WooHeaderPageTitle still follows the h1 format and styling
							<Text
								className={ `woocommerce-layout__header-heading` }
								as="h1"
							>
								{ child }
							</Text>
						</WooHeaderItem>
					);
				} else if ( child.type && child.type.name === 'WooHeaderItem' ) {
                                            const childKey = child.key || childIndex;
				          return cloneElement( child, { key: childKey } );;
				} else {
			            return null;
				}
			} ).filter( child => child !== null );
                }
		</Fill>
	);
}

WooHeader.Slot = ( { fillProps } ) => (
	<Slot name={ 'woocommerce_header_item' } fillProps={ fillProps }>
		{ ( fills ) => {
                         const mergedFills = fills.reduce( ( merged, fill) => {
                            return [...merged, ...fill ];
                         }, []);
		        // could filter out a duplicate PageHeaderTitle here and use a default if none exists
			return fills.sort( ( a, b ) => {
				return a[ 0 ].props.order - b[ 0 ].props.order;
			} );
		} }
	</Slot>
);
```
3PD usage would be:
```
<WooHeader>
  <WooHeaderItem order={-1} ><BackButton /></WooHeaderItem>
  <WooHeaderPageTitle>Title</WooHeaderPageTitle>
</WooHeader>
```

Thoughts? questions?

psealock commented:
>We need to export the header slots for use externally. GB has very specific packages for each area, but this may be overkill for our use case (i.e. @woocommerce/header). Depending on how much we anticipate adding, we could do create a @woocommerce/layout package. Up for discussion.

Awesome to see SlotFill getting some adoption. I haven't fully grasped this PR, but I thought I'd weigh in on this comment because some forethought at this time could ease future headaches.

I do agree `@woocommerce/header` feels overkill, but considering `@woocommerce/navigation` contains the Navigation slotFills, this wouldn't be too different. What does a fully reactified WooCommerce world look like? I imagine there would be hooks in reports, nav, header, product edit, settings, and many more. In this context, `@woocommerce/header` makes sense, especially if Gutenberg behaves the same way. 

In any case, I see the Fill isn't exported in this PR, which is good because it can be done in a follow up PR to isolate this discussion. I hope to take a closer look and weigh in on the other ideas here soon.

joshuatf commented:
>I imagine there would be hooks in reports, nav, header, product edit, settings, and many more. In this context, @woocommerce/header makes sense, especially if Gutenberg behaves the same way.

Great comments

>I have been poking around with this for a little while longer and propose a slightly different slot-fill design, which could potentially go two ways, but will give a cleaner API for the sorting.

💯  love this.  Experimenting with this sorting now.

joshuatf committed:
Allow order prop to header item fill

joshuatf committed:
Split header into before and after

joshuatf committed:
Fix header title gaps

joshuatf committed:
Fix nav and mobile app banner placement

joshuatf commented:
@louwie17 Thanks again for the awesome idea for ordering.  I've incorporated that into the SlotFill in https://github.com/woocommerce/woocommerce-admin/pull/7805/commits/1e9a05dc3618f7d527861e62de637c977fc10de5.

This required a little bit of finessing since SlotFill does not play well with FaCCs which we use to retrieve the `fillProps`.  I traced this back to [here](https://github.com/WordPress/gutenberg/blob/a61153b1ecb43d1efea6441d5c541b415026c3ae/packages/components/src/slot-fill/fill.js#L52-L55) where the props get passed directly to the function.  Unfortunately that meant that FaCC never got called with the `fillProps` if it was wrapped with an HoC and it doesn't seem like there's a way to pass them down further.  I ended up with a solution using `bind` that allows us to get the `order` and the `fillProps`.  I'm open to other options there, but could not seem to get both working without this technique in testing.

I avoided the pattern of wrapping with `<WooHeader>` since this feels compositional in nature like we're creating a new header as opposed to filling an existing slot.  I ended up splitting the header slot by the order (`< 0` goes in the "before" page header area and `>= 0` goes in the "after" page header slot).  AFAIK this pattern hasn't been used before, but I don't foresee any major issues with it.  https://github.com/woocommerce/woocommerce-admin/pull/7805/commits/49d5968451ef07fa8591239c72354dc367021ca9

Let me know what you think.

joshuatf committed:
Fix display options import

joshuatf committed:
Only use last item for page header title fill

joshuatf commented:
@louwie17 I also have a couple failing tests (I believe due to SlotFill not playing well with our test suite). Did you solve this before?

louwie17 commented on client/header/utils.js:
```diff
@@ -0,0 +1,99 @@
+ * @param {Array} param0.children - Node children.
+ * @param {Array} param0.order - Node order.
+ */
+export const WooHeaderItem = ( { children, order = 1 } ) => {
```
I wonder if we should just create a `WooHeaderItemBefore` and `WooHeaderItemAfter` with an order instead of using the negative order values.
The negative vs positive seems a little hacky, and creating two separate HeaderItems might be easier.

joshuatf replied:
I'm a little mixed on this mostly due to the naming.  "Before" and "After" make me think "before/after what?"   The answer to that is `WooHeaderItemBeforePageTitle` but that's quite long and not very pleasing to read.

`WooHeaderItemLeft` might be slightly better, but let me give this some more thought.

joshuatf replied:
>I would steer away from "Left" or "Right", because they are unclear when a right-to-left layout is used (such as Arabic and Hebrew).

Great point, @mattsherman.

>WooHeaderItemBeforePageTitle certainly is wordy, but it is pretty clear.

I agree that it's more clear than Before/After, though I wonder if something with more intent may be beneficial here.

Maybe  `WooHeaderItem` is retained for the header items to the right and `WooHeaderNavigation` is the area to the left which is primarily for navigational use.  Any thoughts on this @mattsherman @louwie17?

louwie17 replied:
> Maybe WooHeaderItem is retained for the header items to the right and WooHeaderNavigation is the area to the left which is primarily for navigational use

I quite like that approach and naming +1 for me.

mattsherman replied:
Oh, I like that... `WooHeaderNavigation` +1

louwie17 commented on client/header/utils.js:
```diff
@@ -0,0 +1,99 @@
+
+	return (
+		<Fill name={ `woocommerce_header_item_${ suffix }` }>
+			{ createOrderedChildren.bind( null, children, order ) }
```
Nice work, I would of never thought about using `bind`. Having said that, so doing the code below doesn't work? I might be missing something.
```
<Fill name={ `woocommerce_header_item_${ suffix }` }>
  { ( fillProps ) => {
      return createOrderedChildren( children, order, {
	 ...fillProps,
	...props,
    } );
  } }
</Fill>
```

joshuatf replied:
I hadn't tried that after creating `createOrderedChildren`, but apparently it does work as long as we are manually cloning the children afterwards.  Update in https://github.com/woocommerce/woocommerce-admin/pull/7805/commits/9ad113680338f96e8a538e74a7ec4f71cc7b6bb1

louwie17 reviewed (commented):
Nice work @joshuatf, thanks for making use of some of my solution, it's looking pretty great!
I just left a couple small comments one in relations to the FaCCS with `fillProps`.  But outside of those, I say this is pretty well ready.

> I avoided the pattern of wrapping with <WooHeader> since this feels compositional in nature like we're creating a new header as opposed to filling an existing slot.

Yeah that makes sense, good decision! 

> I ended up splitting the header slot by the order (< 0 goes in the "before" page header area and >= 0 goes in the "after" page header slot). AFAIK this pattern hasn't been used before, but I don't foresee any major issues with it.

I just left a comment, I would be in favour of just creating two separate fill components a `Before` & `After`.

> @louwie17 I also have a couple failing tests (I believe due to SlotFill not playing well with our test suite). Did you solve this before?

Originally I thought I solved this, but I never did get it working. I end up just mocking out the fill components we export. I am not a 100% sure why the slot-fills don't play nice.
It seemed to work well for me if I used slot-fill manually in tests, but as soon as I tried using our wrapper components it broke. Something to do with the Provider I think.

joshuatf committed:
Use function to pass fill props instead of bind

joshuatf committed:
Rename header slots

joshuatf commented:
Thanks for all the discussion @louwie17 @mattsherman!  I've addressed the feedback here and updated the SlotFill naming.

I have not renamed the one ActivityPanels to ActivityWidgets.  I started doing this, but it creates quite a big diff and I think may be pushing the scope on this PR a bit.  If it's alright, I think we'll save for a follow-up given this is a pre-existing issue.

louwie17 commented on client/mobile-banner/index.js:
```diff
@@ -1,93 +1,37 @@
-			</div>
-		);
+	if ( ! isDismissed ) {
+		return null;
```
This should be `if ( isDismissed ) {` right? not `! isDismissed`?

joshuatf replied:
Definitely!  Good catch @louwie17 ; updated in https://github.com/woocommerce/woocommerce-admin/pull/7805/commits/5c054521b7878bbedc3908aad4d1a348fbfe4b1e

louwie17 reviewed (commented):
@joshuatf thanks for sticking around with all this feedback, things are looking great!!
I just noticed one minor issue with the Android banner logic, outside of that it's ready to approve (will wait for the above to be fixed first though).

joshuatf committed:
Fix mobile banner dismissal check

louwie17 reviewed (approved):
Thanks @joshuatf, approving this then, only the unit tests and E2E tests need to be updated, I think just mocking out the slot fills for the most part will be the best solution there.
Anyway approving this as those things aren't blockers in this case.

joshuatf committed:
Fix up inbox panel rename

joshuatf committed:
Update task title in tests

joshuatf committed:
Fix up task status retrieval

joshuatf commented:
@louwie17 Turns out we didn't need to mock those slots at all since we directly test the component.  The test failures were due to random rebase conflicts that should all be fixed up now. 🤞 

Would you mind giving this another look when you can just to sanity check the changes before I merge?

louwie17 commented:
> @louwie17 Turns out we didn't need to mock those slots at all since we directly test the component. The test failures were due to random rebase conflicts that should all be fixed up now. 🤞
> 
> Would you mind giving this another look when you can just to sanity check the changes before I merge?

Looks good @joshuatf, glad the test failures where just merge conflicts, glad to get this over the line now 👍
